src/Bonami/Collection/Monoid/Monoid.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bonami\Collection\Monoid;
+
+/** @phpstan-template A */
+interface Monoid
```

Comment:
Did we consider abstract class instead of interface?

we could disallow constructor there & add some factory + combinators...

But it can be changed anytime in the future...

Reply:
I would not force creating Monoids via general factories. Imagine you would have structure for Price, which has mandatory Currency and cannot exists as empty without that currency. Then you would need Monoid like this:

```php
class PriceMonoid implements Monoid
{
    private $currency;

    public function __construct(Currency $currency)
    {
        $this->currency = $currency;
    }

    public function concat($a, $b)
    {
        return $a->sum($b);
    }

    public function getEmpty()
    {
        new Price(new BigInt(0), $this->currency);
    }
}
```

This will let you more open for implementations like this.

BTW this is real world example. I know, that it would be better to have reprezentation of "empty" Price as Price without currency at all, but sometime you need to deal with existing classes like this one.

Reply:
Makes sense, but what about combinators? Put them somewhere as a standalone functions? I have no issue with that, but we dont use that style in project lot (except for stuff in `helpers.php`).

Reply:
We will explore it later, I think that this does not close door for us for combinators.